compare_platforms/include/review.h

```diff
@@ -0,0 +1,151 @@
+#endif
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * Reviewer's username, real name, private comments, public review, and permissions. 
+**/
+struct review
+{
+    char *username;
+    char *realname;
+    char *privatereview;
+    char *publicreview;
+    char *permissions;
+};
+
+/**
+ * Check this review's permissions.
+ * param: review, the struct with the details of the review whose permissions we are checking
+ * return: true if the review has write permissions, false otherwise.
+**/
+bool can_write(struct review *rv) {
+    bool canwrite = true;
+    
+    // a rigorous protocol of looking for the letter w in the permissions
+    char *pt = strchr(rv->permissions, 'w');
+    if(pt == NULL)
+        return false;
+
+    return canwrite;
+}
+
+/**
+ * A method for editing the member of the review structure that holds a public review.
+ * param: rv, the struct representing the review in question
+ * param: publicreview, a string with the public review intended to replace the existing one
+ * return: false if review is read-only, true if review is writable.
+**/
+bool change_publicreview(struct review *rv, char* publicreview)
+{
+    #ifdef __CHERI_PURE_CAPABILITY__
+        // If this is read only, it will crash.
+        // To test this in a CHERI environment,
+        // we won't do a weak check of read-only permissions here.
+        rv->publicreview = publicreview;
+        // If we got this far, it was writable.
+        return true;
+    #else
+        if(can_write(rv)) {
+            rv->publicreview = publicreview;
+            return true;
+        }
+        else {
+            printf("You don't have permission to change this public review.\n");
+            return false;
+        }
+    #endif
+}
+
+/**
+ * A method for attempting to change the public review to something more flattering.
+ * param: rv, the struct representing the public review in question.
+ * param: reqsz, a size_t representing the amount of memory to allocate for the new review.
+ * param: b_weak, a boolean that is true if we are weakly enforcing (i.e. using can_write) the 
+ * read-only permissions, false otherwise.
+ * return: false if review is read-only, true if review is writable.
+**/
+bool try_improve_publicreview(struct review *rv, size_t reqsz, bool b_weak) {
+    bool b_writable = false;
+    
+    char *newpublicreview = malloc(reqsz);
+    strcpy(newpublicreview, "5 ***** strong accept. This author deserves two Nobels and an ice cream.\n");
+
+    printf("\nAttempting to change the public review.\n");
+    #ifdef __CHERI_PURE_CAPABILITY__
+        if((can_write(rv) == false) && (b_weak == true)) {
+            printf("You do not have permission to change the public review.\n");
+            return false;
+        } else {
+            // if the review is set to read-only, this will crash.
+            b_writable = change_publicreview(rv, newpublicreview);
```

Comment:
I think this `ifdef` here needs re-organizing. There is one in `change_publicreview` as well. Could we reduce the number of places we check for CHERI minimally, so we can say "aha, here's the exact place where divergence happens"?

Reply:
Done. CHERI checks are carried out in two places now: inside set_read_only, and in main().

Reply:
> I think this `ifdef` here needs re-organizing. There is one in `change_publicreview` as well. Could we reduce the number of places we check for CHERI minimally, so we can say "aha, here's the exact place where divergence happens"?

Fixed in ccb5c32572367fd59362ebc3b64e557b811a7962